### PR TITLE
#fixed information schema crash 883

### PR DESCRIPTION
**Pull request by jamesstout**
## Changes:
In MySQL v8 or above, the `information_schema` 'table' is now a _view_. So the parsing code used in `informationForTable` no longer works. This change uses `SHOW COLUMNS` in place of `SHOW CREATE TABLE` and parses those results. 

From the [docs](https://dev.mysql.com/doc/refman/8.0/en/show-columns.html):

> `SHOW COLUMNS` displays information about the columns in a given table. **It also works for views**. 

The output is the same for < 8.0.0 and > 8.0.0 so we can change to this one method.

**IT NEEDS TESTING - hence only a draft PR**

## Closes following issues:
- #833  

## Tested:
- Processors:
  - [x] Intel
  - [ ] Apple Silicon
- macOS Versions:
  - [ ] 10.12.x (Sierra)
  - [ ] 10.13.x (High Sierra)
  - [ ] 10.14.x (Mojave)
  - [x] 10.15.x (Catalina)
  - [ ] 11.x (Big Sur)
- Xcode Version: 12.4 (12D4e)
- Localizations:
  - [ ] English
  - [ ] Spanish
  - [ ] Other (please specify)
## Additional notes:
For v8+ we lose a bit of extended table info, shown on the Table Info screen, as this data is no longer available. e.g. size, max data len.

### CODE REVIEW

File: Source/Controllers/DataControllers/SPTableData.m [hunk 1]
```diff
@@ -516,9 +541,90 @@ - (NSDictionary *) informationForTable:(NSString *)tableName fromDatabase:(NSStr
...
+    SPMySQLResult *theResult;
+    NSString *queryStr;
+
+    SPLog(@"createTableSyntaxFromView");
+
+    queryStr = [NSString stringWithFormat:@"SHOW COLUMNS FROM %@", [tableName backtickQuotedString]];
+
+    SPLog(@"queryStr: %@", queryStr);
+
+    theResult = [mySQLConnection queryString:queryStr];
```
Comment (Kaspik) [inline Source/Controllers/DataControllers/SPTableData.m:560]: Why not to define the `theResult` here?

File: Source/Controllers/DataControllers/SPTableData.m [hunk 2]
```diff
@@ -516,9 +541,90 @@ - (NSDictionary *) informationForTable:(NSString *)tableName fromDatabase:(NSStr
...
+
+    NSDictionary *tableRow;
+    while ((tableRow = [theResult getRowAsDictionary]) != nil) {
+
+        NSString *Default = [tableRow safeObjectForKey:@"Default"];
+        NSString *Field = [tableRow safeObjectForKey:@"Field"];
+        NSString *Null = [tableRow safeObjectForKey:@"Null"];
+        NSString *Type = [tableRow safeObjectForKey:@"Type"];
+
+        // FIXME: are we sure this handles all scenarios?
```
Comment (Kaspik) [inline Source/Controllers/DataControllers/SPTableData.m:576]: no idea

File: Source/Controllers/DataControllers/SPTableData.m [hunk 3]
```diff
@@ -1004,19 +1110,15 @@ - (BOOL)updateStatusInformationForCurrentTable
 
 		// Reassign any "Type" key - for MySQL < 4.1 - to "Engine" for consistency.
```
Comment (Kaspik) [inline Source/Controllers/DataControllers/SPTableData.m:1111]: This should be... removed? Or all these workarounds?

File: Source/Controllers/DataControllers/SPTableData.m [hunk 4]
```diff
@@ -1004,19 +1110,15 @@ - (BOOL)updateStatusInformationForCurrentTable
...
 		if ([status objectForKey:@"Type"]) {
-			[status setObject:[status objectForKey:@"Type"] forKey:@"Engine"];
+			[status safeSetObject:[status objectForKey:@"Type"] forKey:@"Engine"];
 		}
 
 		// If the "Engine" key is NULL, a problem occurred when retrieving the table information.
```
Comment (Kaspik) [inline Source/Controllers/DataControllers/SPTableData.m:1116]: Remove these comments then?

File: Source/Controllers/DataControllers/SPTableData.m [hunk 5]
```diff
@@ -1028,26 +1130,31 @@ - (BOOL)updateStatusInformationForCurrentTable
 			tableStatusResult = [mySQLConnection queryString:[NSString stringWithFormat:@"SELECT COUNT(1) FROM %@", [escapedTableName backtickQuotedString] ]];
 			// this query can fail e.g. if a table is damaged
 			if (tableStatusResult && ![mySQLConnection queryErrored]) {
-				[status setObject:[[tableStatusResult getRowAsArray] objectAtIndex:0] forKey:@"Rows"];
+				[status safeSetObject:[[tableStatusResult getRowAsArray] safeObjectAtIndex:0] forKey:@"Rows"];
```
Comment (Kaspik) [inline Source/Controllers/DataControllers/SPTableData.m:1133]: If you stop using objectAtIndex for index 0, but start using `firstObject`, you will save a lot of time for yourself :)

File: Source/Controllers/DataControllers/SPTableData.m (hunk 3 above)
Comment (jamesstout) [reply Source/Controllers/DataControllers/SPTableData.m:1111]: don't know. doesn't hurt to leave them?

File: Source/Controllers/DataControllers/SPTableData.m (hunk 4 above)
Comment (jamesstout) [reply Source/Controllers/DataControllers/SPTableData.m:1116]: OK

File: Source/Controllers/DataControllers/SPTableData.m (hunk 5 above)
Comment (jamesstout) [reply Source/Controllers/DataControllers/SPTableData.m:1133]: It's not me using ... it's existing code! But yes, I should check if index is 0 when fixing these crashes.

File: Source/Controllers/DataControllers/SPTableData.m (hunk 2 above)
Comment (jamesstout) [reply Source/Controllers/DataControllers/SPTableData.m:576]: hence testing needed :)

File: Source/Controllers/DataControllers/SPTableData.m (hunk 5 above)
Comment (jamesstout) [reply Source/Controllers/DataControllers/SPTableData.m:1133]: so go ahead and change them all?

Comment (Kaspik) [reply Source/Controllers/DataControllers/SPTableData.m:1133]: Don't have to do it all, but `objectAtIndex:0` should be replaced for safety... up to you :)

File: Source/Other/Data/SPConstants.h [hunk 6]
```diff
@@ -258,6 +258,7 @@ extern NSString *SPMySQLDatabase;
 extern NSString *SPMySQLInformationSchemaDatabase;
 extern NSString *SPMySQLPerformanceSchemaDatabase;
 extern NSString *SPMySQLSysDatabase;
+extern NSString *SPMySQLTempTableName;
```
Comment (Jason-Morcos) [inline Source/Other/Data/SPConstants.h:261]: Eliminate this constant?

File: Source/Other/Data/SPConstants.m [hunk 7]
```diff
@@ -64,6 +64,7 @@ - (SPTableViewType)tableViewTypeEnumFromString{
 NSString *SPMySQLInformationSchemaDatabase       = @"information_schema";
 NSString *SPMySQLPerformanceSchemaDatabase       = @"performance_schema";
 NSString *SPMySQLSysDatabase                     = @"sys";
+NSString *SPMySQLTempTableName                   = @"my_temp_table";
```
Comment (Jason-Morcos) [inline Source/Other/Data/SPConstants.m:67]: Eliminate this constant?

File: Source/Controllers/DataControllers/SPTableData.m [hunk 8]
```diff
@@ -464,10 +466,19 @@ - (NSDictionary *) informationForTable:(NSString *)tableName fromDatabase:(NSStr
...
     SPMySQLResult *theResult;
-    if (database)
-        theResult = [mySQLConnection queryString:[NSString stringWithFormat:@"SHOW CREATE TABLE %@.%@", [database backtickQuotedString], [tableName backtickQuotedString]]];
-    else
-        theResult = [mySQLConnection queryString:[NSString stringWithFormat:@"SHOW CREATE TABLE %@", [tableName backtickQuotedString]]];
+    NSString *queryStr;
+
+    if (database){
+        queryStr = [NSString stringWithFormat:@"SHOW CREATE TABLE %@.%@", [database backtickQuotedString], [tableName backtickQuotedString]];
+        theResult = [mySQLConnection queryString:queryStr];
```
Comment (Jason-Morcos) [inline Source/Controllers/DataControllers/SPTableData.m:473]: Could move theResult to outside/after the IF statement as it's the same for both the IF and ELSE conditions

File: Source/Controllers/DataControllers/SPTableData.m [hunk 9]
```diff
@@ -498,9 +509,23 @@ - (NSDictionary *) informationForTable:(NSString *)tableName fromDatabase:(NSStr
     NSArray *syntaxResult = [theResult getRowAsArray];
     NSArray *resultFieldNames = [theResult fieldNames];
 
+    SPLog(@"syntaxResult: %@", syntaxResult);
+    SPLog(@"resultFieldNames: %@", resultFieldNames);
+
+    // MARK: removed isMySQL8 == YES, so this is used for all versions
+    if([database isEqualToString:SPMySQLInformationSchemaDatabase]){
```
Comment (Jason-Morcos) [inline Source/Controllers/DataControllers/SPTableData.m:516]: Is this the best check we should use to determine when to use the view logic? Should be perhaps check the structure of the response instead?

File: Source/Controllers/DataControllers/SPTableData.m [hunk 10]
```diff
@@ -516,9 +541,87 @@ - (NSDictionary *) informationForTable:(NSString *)tableName fromDatabase:(NSStr
...
+
+    NSString *queryStr = [NSString stringWithFormat:@"SHOW COLUMNS FROM %@", [tableName backtickQuotedString]];
+
+    SPLog(@"queryStr: %@", queryStr);
+
+    SPMySQLResult *theResult = [mySQLConnection queryString:queryStr];
+
+    NSMutableString *viewCreateStr = [[NSMutableString alloc] init];
+
+    [viewCreateStr appendFormat:@"CREATE TEMPORARY TABLE `%@` (\n", tableName];
```
Comment (Jason-Morcos) [inline Source/Controllers/DataControllers/SPTableData.m:561]: Will this require special permissions to the database in order to do? In other words, can I create a temporary table even from a readonly connection?

**Comment by Kaspik** [comment]
How does it look here with this PR, tested, verified? :)

**Comment by jamesstout** [comment]
> How does it look here with this PR, tested, verified? :)

Well I've tested it on 5.6, 5.7, 8.0.22 and MariaDB v10.4.17. Seems OK.

File: Source/Controllers/DataControllers/SPTableData.m (hunk 2 above)
Comment (Kaspik) [reply Source/Controllers/DataControllers/SPTableData.m:573]: Soooo, can we get rid of the fixme? :)

**Comment by Jason-Morcos** [comment]
@Kaspik Do you want to create 3.2.0 rc 1 and submit including this PR? I can too, just let me know

**Review by Jason-Morcos — APPROVED** [review]

**Comment by Kaspik** [comment]
Yeah I can do it tomorrow if you don't do it today :)

**Pull request merged**